editor/src/clj/editor/process.clj

```diff
@@ -13,31 +13,157 @@
+            (values) or nil (removes the var from the process env)"
+  ^Process [& opts+args]
+  (let [has-opts (map? (first opts+args))
+        opts (if has-opts (first opts+args) {})
+        command (if has-opts (rest opts+args) opts+args)
+        {:keys [in out err dir env]} opts
+        pb (ProcessBuilder. ^List command)]
+    (when dir (.directory pb (io/file dir)))
+    (when in (.redirectInput pb (->redirect in)))
+    (when out (.redirectOutput pb (->redirect out)))
+    (when err
+      (if (= :stdout err)
+        (.redirectErrorStream pb true)
+        (.redirectError pb (->redirect err))))
     (when env
-      (let [environment (.environment pb)]
+      (let [m (.environment pb)]
         (doseq [[k v] env]
-          (.put environment k v))))
-    (when removeenv
-      (let [environment (.environment pb)]
-        (doseq [k removeenv]
-          (.remove environment k))))
-    (when (some? redirect-error-stream?)
-      (.redirectErrorStream pb (boolean redirect-error-stream?)))
+          (if v
+            (.put m k v)
+            (.remove m k)))))
     (.start pb)))
 
-(defn watchdog! ^Thread [^Process proc on-exit]
-  (doto (Thread. (fn []
-                   (.waitFor proc)
-                   (on-exit)))
-    (.start)))
+(defn out
+  "Get the process out, an InputStream that can be read from"
+  ^InputStream [^Process process]
+  (.getInputStream process))
+
+(defn in
+  "Get the process in, an OutputStream that can be written to"
+  ^OutputStream [^Process process]
+  (.getOutputStream process))
+
+(defn err
+  "Get the process err, an InputStream that can be read from"
+  ^InputStream [^Process process]
+  (.getErrorStream process))
+
+(defn ^{:arglists '([^InputStream in & {:keys [buffer-size encoding] :or {buffer-size 1024 encoding "UTF-8"}}])}
+  capture
+  "Given an InputStream such as process's out or err, consume it into a string
+
+  Returns a non-empty string with trimmed new lines or nil if it's empty
+
+  Optional kv-args:
+    :buffer-size    read buffer size, default 1024
+    :encoding       text encoding, default UTF-8"
+  [^InputStream input-stream & opts]
+  (let [w (StringWriter.)]
+    (apply io/copy input-stream w opts)
+    (let [s (string/trim-newline (.toString w))]
+      (when (pos? (.length s))
+        s))))
+
+(defn pipe
+  "Pipe data from the InputStream (e.g. [[out]] or [[err]]) to OutputStream
+
+  Similar to io/copy and InputStream::transferTo, but flushes on write"
+  [^InputStream in ^OutputStream out]
+  (let [buf (byte-array 1024)]
+    (loop []
+      (let [size (.read in buf)]
+        (when-not (neg? size)
+          (do (.write out buf 0 size)
+              (.flush out)
+              (recur)))))))
+
+(defn await-exit-code
+  "Blocks until process exits and returns the exit code"
+  [^Process process]
+  (.waitFor process))
+
+(defn exit-ok?
+  "Checks if the exit code of the process indicates success"
+  [exit-code]
+  {:pre [(int? exit-code)]}
+  (zero? exit-code))
```

Comment:
I don't think we need this function. It sort of feels like it should take a `Process` which is confusing. We can use `zero?` with `await-exit-code` directly whenever this would have been used.

Reply:
Yeah, that was unnecessary.